File: 백준/스택/송예은/오큰수.py
#풀이 방법이 같은데 틀린 이유 모르겠음

N = int(input())
stack =[]
ans = []

numbers = list(map(int,input().split()))

for i in range(len(numbers)-1,-1,-1):
    print(stack,ans)
    if len(stack) == 0:
        stack.append([numbers[i],i+1])
        ans.append(-1)
    else :
        while len(stack) > 0:
            if stack[-1][0] < numbers[i]:
                stack.pop()
            else :
                break

        stack.append([numbers[i],i+1])

        if len(stack) == 1:
            ans.append(-1)
        else:
            ans.append(stack[-2][0])
print(*ans[::-1])

File: 백준/스택/문종후/오큰수.js
let [N, ...input] = require("fs").readFileSync("dev/stdin").toString().trim().split("\n");

function solution(N, input) {
  let arr = input[0].split(" ").map(Number);
  let answer = Array.from({ length: N }, () => -1);
  let stack = [];
  for (let i = 0; i < N; i++) {
    while (stack.length && arr[stack[stack.length - 1]] < arr[i]) {
      answer[stack.pop()] = arr[i];
    }
    stack.push(i);
  }
  return answer.join(" ");
}
console.log(solution(N, input));
